src/meta/src/model/system_param.rs

```diff
@@ -0,0 +1,189 @@
+// Copyright 2023 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::HashSet;
+
+use async_trait::async_trait;
+use risingwave_pb::meta::SystemParams;
+
+use super::{MetadataModel, MetadataModelError, MetadataModelResult};
+use crate::storage::{MetaStore, Snapshot, Transaction};
+
+const SYSTEM_PARAM_CF_NAME: &str = "cf/system_params";
+
+const BARRIER_INTERVAL_MS_KEY: &str = "barrier_interval_ms";
+const CHECKPOINT_FREQUENCY_KEY: &str = "checkpoint_interval";
+const SSTABLE_SIZE_MB_KEY: &str = "sstable_size_mb";
+const BLOCK_SIZE_KB_KEY: &str = "block_size_kb";
+const BLOOM_FALSE_POSITIVE_KEY: &str = "bloom_false_positive";
+const STATE_STORE_KEY: &str = "state_store";
+const DATA_DIRECTORY_KEY: &str = "data_directory";
+const BACKUP_STORAGE_URL_KEY: &str = "backup_storage_url";
+const BACKUP_STORAGE_DIRECTORY_KEY: &str = "backup_storage_directory";
+
+#[async_trait]
+impl MetadataModel for SystemParams {
+    type KeyType = u32;
+    type ProstType = SystemParams;
+
+    fn cf_name() -> String {
+        SYSTEM_PARAM_CF_NAME.to_string()
+    }
+
+    fn to_protobuf(&self) -> Self::ProstType {
+        // System params are stored as raw kv pairs.
+        unimplemented!()
+    }
+
+    fn from_protobuf(_prost: Self::ProstType) -> Self {
+        // System params are stored as raw kv pairs.
+        unimplemented!()
+    }
+
+    fn key(&self) -> MetadataModelResult<Self::KeyType> {
+        // System params are stored as raw kv pairs.
+        unimplemented!()
+    }
```

Comment:
```suggestion
    fn to_protobuf(&self) -> Self::ProstType {
        // System params are stored as raw kv pairs.
        unreachable!()
    }

    fn from_protobuf(_prost: Self::ProstType) -> Self {
        // System params are stored as raw kv pairs.
        unreachable!()
    }

    fn key(&self) -> MetadataModelResult<Self::KeyType> {
        // System params are stored as raw kv pairs.
        unreachable!()
    }
```

Reply:
Hmmm, actually `to_protobuf/from_protobuf` are not `unimplemented`, but just `self`? And `key` is string? 🤔 cc @yezizp2012 for review

Reply:
I think the indication of the trait `MetadataModel` is that whatever struct that implements it correspond to **a single** KV pair in meta store. Since we want to store them as multiple readable KV pairs https://github.com/risingwavelabs/rfcs/pull/34#issuecomment-1415117338, proto format should be banned.

Reply:
If so, I think it's better to directly impl some necessary functions for `SystemParams` rather than the trait `MetadataModel`.